Allow running subsets of tests [QUE-402]

stoeffel opened the pull request:
Allows running only a subset of tests by passing filepaths with line numbers

jwoudenberg-nri commented on nri-log-explorer/package.yaml:
```diff
@@ -22,11 +22,14 @@ executables:
     - brick >= 0.51 && < 1.0
     - bytestring >= 0.10.8.2 && < 0.12
     - directory >= 1.3.3.0 && < 1.4
+    - fuzzy >= 0.1.0.0 && < 0.2 
```
💭 This seems to pull in a big dependency `monoid-subclasses`. Might be worth seeing if we can drop it at some point.

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -42,15 +45,18 @@ import qualified System.Exit
     -- A tool like pbcopy or xclip for copying to clipboard.
     clipboardCommand :: Maybe Text,
+    -- All root spans.
+    allRootSpans :: List RootSpan,
```
❔ Given that `ListWidget.List` is backed by a `Vector`, I wonder if keeping the original list in a `Vector` as well will mean we need to do fewer conversions when updating a filter.

stoeffel replied:
done

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -59,9 +65,15 @@ data Model = Model
+data Filter
+  = NoFilter
+  | HasFilter Text (List Text)
+  | Filtering (Maybe (Text, List Text)) (Edit.Editor Text Name)
```
😻

jwoudenberg-nri replied:
🎨 Maybe rename `Filtering` to `EditingFilter`?

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -82,6 +94,11 @@ data Msg
   | CopyDetails
+  | ShowFilter
+  | ClearFilter
+  | StopFiltering (Maybe (Text, List Text))
```
🎨 Maybe `StopEditingFilter`?

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -90,34 +107,35 @@ data Msg
-  = NoDataPage
-  | RootSpanPage Time.UTCTime (ListWidget.List Name RootSpan)
+  = NoDataPage Filter
+  | RootSpanPage Time.UTCTime Filter (ListWidget.List Name RootSpan)
```
😻

stoeffel commented:
@jwoudenberg-nri I've addressed all your comments.
I think there are no alternative packages for fuzzy that I can find on hackage :-(
I think our options are drop fuzzy filtering (it's so nice though) or write/fork our own fuzzy. I'm happy to fork, but I think that's out of scope for this PR, except if you think it's blocking it. Thoughts?

jwoudenberg-nri reviewed (approved):
This is great, thank you so much!

![](https://media.giphy.com/media/4QFAV6hISzSjPyaWS2/giphy.gif)

jwoudenberg-nri commented:
> @jwoudenberg-nri I've addressed all your comments.

Sorry, was still reviewing so I added a few more!

> I think there are no alternative packages for fuzzy that I can find on hackage :-(
> I think our options are drop fuzzy filtering (it's so nice though) or write/fork our own fuzzy. I'm happy to fork, but I think that's out of scope for this PR, except if you think it's blocking it. Thoughts?

I definitely want to keep fuzzy searching. I wonder if creating our own function for matching would be a lot of work, but definitely agree it's out of scope for this PR. Just a thought that occurred to me while reviewing, nothing I feel we need to act on now!

stoeffel commented on nri-log-explorer/src/Main.hs:
```diff
@@ -198,6 +215,75 @@ update model msg =
+                    EditFilter previous _ -> EditFilter previous editor
+                    HasFilter first rest ->
+                      EditFilter (Just (first, rest))
+                        <| Edit.applyEdit (\_ -> TZ.gotoEOL <| TZ.textZipper [Text.join " " (first : rest)] (Just 1)) editor
```
We do get live updates as we type.
Might make it a little simpler though. will check

stoeffel replied:
Actually, I think it makes it harder. Because we also want to track previous filters for the esc behaviour.

stoeffel commented on nri-log-explorer/src/Main.hs:
```diff
@@ -530,36 +698,55 @@ handleEvent ::
-            (ListWidget.handleListEventVi ListWidget.handleListEvent vtyEvent)
-            model
-            |> andThen Brick.continue
+      case filter model of
```
I think we can't call `handleEditorEvent` in "update" though right?

stoeffel committed:
Run should take a list of filepaths to run

stoeffel committed:
Add filtering of subset

stoeffel committed:
Run only tests that match the srcLoc

stoeffel committed:
Update golden results

stoeffel committed:
Add test for running subset

stoeffel committed:
Run only one file without line numbers

stoeffel committed:
REad files from args

stoeffel committed:
Allow srcloc to be suffix

omnibs commented on nri-prelude/src/Test/Internal.hs:
```diff
@@ -390,8 +391,14 @@ run runSubset (Test all) = do
+subset filePaths singleTest =
+  case (filePaths, loc singleTest) of
+    ([], _) -> False
+    (_, Nothing) -> False
```
Do you know when we can have a test w/o a corresponding `SrcLoc`?

omnibs replied:
We discussed it and it seems it might not happen in practice. Maybe we should remove the `Maybe` from `loc`.

omnibs commented on nri-prelude/src/Test/Internal.hs:
```diff
@@ -401,7 +401,7 @@ subset filePaths singleTest =
               then ":" ++ Prelude.show srcLocStartLine
               else ""
-       in if x == srcLocFile ++ postfix
+       in if (srcLocFile ++ postfix) `isSuffixOf` x
```
Explain what this does?

omnibs committed:
Small explanation

omnibs committed:
Support passing any LoC inside test body

The splitting logic will be moved to arg parsing so we can fail in a
nicer way when the args are bogus

omnibs committed:
Write a parser for CLI args

omnibs committed:
Add some tests

omnibs committed:
Improve error messages

omnibs committed:
Handle invalid LoC better, further improve errors

omnibs committed:
Use the parsed args for running subsets of tests

omnibs committed:
Simplify parsing failure and fix some lints

omnibs reviewed (approved):
This looks good, but I made significant changes, so I'd love a second look from @stoeffel before merging

stoeffel committed:
Allow multiple spaces after --files

stoeffel committed:
Remove unnecessary ()

stoeffel committed:
We want the test to show where it failed not in the helper

stoeffel committed:
Cleanup parser and allow trailing ,

stoeffel committed:
Use Result not Either

omnibs committed:
Make LoC for tests not a Maybe, fail early
